Various additional updates for 2018 analysis

cschnaible opened the pull request:
This PR includes changes for 2018 analysis.
- Move SimpleNtuple python cff into its own file
- Add dyt track to SimpleNtuple
- Update MCSamples with 2018 MC (incomplete)
- Loosen cuts on DileptonPreselector 
- Update hltTriggerMatch with 2018 prescaled trigger paths
- nminus1effs.py update for 2018
- Rewrite Efficiency & Resolution histos.py 
- Update Efficiency histos with 2018 trigger OR
- Update job submission scripts for better ease of use
- other cosmetic changes (indention, mostly)

cschnaible committed:
Add prefer_Z=True to OurSelection2018_cff

cschnaible committed:
Update MCSamples with 2018 102X MC where they exist

cschnaible committed:
add calculation of effective number of entries in case of negatively weighted evts to MCSamples, add lep_dyt_* to SimpleNtuple_miniAOD

cschnaible committed:
Bring nminus1effs.py up to date

cschnaible committed:
Update PrescaleToCommon class, hltTriggerMatch_cfi, and histos.py so that computing Z0 cross section is now possible in 2018. Update histos.py and add config file NtupleFromPAT_cfi so that an ntuple can be made for different paths - namely so that an Our2018MuonsPlusMuonsMinusNtuple can exist. Other cleaning too like preventing empty dielectron histograms from being made when you dont want them.

cschnaible committed:
Update EfficiencyResolutionFromMC for 2018 analysis

cschnaible committed:
Update goodlumis for full 2018 good run list

cschnaible committed:
A couple of quick fixes for various submission scripts

cschnaible commented:
Hi All,
I went through the code my mini-AOD-2018 branch and updated it so that it can run all parts of the analysis. I opened a "draft" pull request because, as always, it needs to be tested by someone other than myself. (This "draft" pull request is a new feature in GitHub. If it ends up being more confusing than it is helpful we can probably do without it.)

BenjaminRS commented on plugins/PrescaleToCommon_miniAOD.cc:
```diff
@@ -100,12 +100,12 @@ bool PrescaleToCommon_miniAOD::filter(edm::Event& event, const edm::EventSetup&
-    const std::vector<std::string>& pathList = hlt_cfg.triggerNames();
-    std::cout<<"path size "<<pathList.size()<<std::endl;
+    //const std::vector<std::string>& pathList = hlt_cfg.triggerNames();
+    //std::cout<<"path size "<<pathList.size()<<std::endl;
```
Can we have these as flags for extra debugging information, or is the information provided by the list not relevant anymore?

cschnaible replied:
Judging from the commit history of this file (the non-MiniAOD version), the stl output from this file has been in flux and is likely used mainly for development/debugging purposes. A flag for the stl output with the default value set to False is probably better than just commenting the lines out (it makes it clear for future users what the output's purpose is).

BenjaminRS replied:
Yes I would agree to have it go to a debugging stream with the debugging set to False by default (this is what I meant earlier). I see you have now implemented it. Thanks!

BenjaminRS commented on plugins/CheckPrescale.cc:
```diff
@@ -32,7 +32,8 @@ class CheckPrescale : public edm::EDAnalyzer {
   };
 
-  tree_t last_t, t;
+  //tree_t last_t, t;
```
Is last_t no longer used?

cschnaible replied:
As far as I can tell from the commit history of CheckPrescale.cc last_t was used in 2011. That is, the part of the code where it is actually used was commented out in 2011. All I did was comment out some of the remaining vestigial lines of code.

BenjaminRS commented on python/hltTriggerMatch_cfi.py:
```diff
@@ -25,23 +25,114 @@
+prescaled_trigger_paths = ['HLT_Mu27_v%i' % i for i in (12,13)]
+trigger_paths = ['HLT_Mu50_v%i' % i for i in (12,13)]
+#trigger_paths += ['HLT_OldMu100_v%i' % i for i in (3)]
+#trigger_paths += ['HLT_TkMu100_v%i' % i for i in (2)]
```
doesn't this (and also OldMu100) need to be uncommented?

cschnaible replied:
The trigger_paths list for the primary analysis triggers is not actually used in the trigger code anymore. It has been superseded by the lists trigger_filters and trigger_path_names. However, its equivalent prescaled_trigger_paths is used by PrescaleToCommon so it is updated and kept. Also any code that would use trigger_paths is not set up to handle different trigger paths, only different versions of the same single path. The additional commented info from the hlt summary webpages is just for bookkeeping purposes.

BenjaminRS replied:
Is there any critical code that uses trigger_paths?

cschnaible replied:
The lepton producer (Zprime2muLeptonProducer_miniAOD) uses it as a backup in case trigger_filters or trigger_path_names are not set, but I don't think anything else uses it.

BenjaminRS reviewed (commented):
Some initial comments. @khaosmos93 can you do a proper review and also test the new code as well please.

cschnaible committed:
Add flag for stl debug output

khaosmos93 commented:
Thanks a lot for the updates! I have tested "test/DataMCSpectraComparison/histos.py" and "test/EfficiencyResolutionFromMC/histos.py" so far. I think some parts need to be updated before I test further.

test/DataMCSpectraComparison/histos.py:
1. It seems like "Electron" should be "Electrons" in [1].
2. Have you checked generator filter part [2] is properly working? For me it looks like it applies gen filter for "dyInclusive50" (tautau) to every samples... Otherwise it runs fine.

test/EfficiencyResolutionFromMC/histos.py:
1.  one bug found [3] which looks like just a typo
`extra = '_'+ if ex!='' else ''` -> `extra = '_'+ex if ex!='' else ''`
2. I think current setup might consumes memory too much (RSS ~ 3391 MB), i.e. too large to submit the CRAB jobs... It is likely due to having too many histograms, so in principle one can reduce number of modules (EDAnalyzers) if CRAB fails.

Please correct me if I wrote something wrong.

Thanks

[1] https://github.com/cschnaible/SUSYBSMAnalysis-Zprime2muAnalysis/blob/mini-AOD-2018/test/DataMCSpectraComparison/histos.py#L399
[2] https://github.com/cschnaible/SUSYBSMAnalysis-Zprime2muAnalysis/blob/mini-AOD-2018/test/DataMCSpectraComparison/histos.py#L242-L308
[3] https://github.com/cschnaible/SUSYBSMAnalysis-Zprime2muAnalysis/blob/mini-AOD-2018/test/EfficiencyResolutionFromMC/histos.py#L216

cschnaible committed:
Adding comments and bug fixes for DataMCSpectraComparison/histos.py EfficiencyResolutionFromMC/histos.py

cschnaible commented:
Hi Min, Thanks for taking a look.

> test/DataMCSpectraComparison/histos.py:
> 
> 1. It seems like "Electron" should be "Electrons" in [1].
> 2. Have you checked generator filter part [2] is properly working? For me it looks like it applies gen filter for "dyInclusive50" (tautau) to every samples... Otherwise it runs fine.

No, I haven't checked if any of the generator filters are properly working, but they're simple enough that I doubt they will cause problems. 
Also, I've never been happy with how the generator filters are implemented because it requires some amount of manual intervention to turn them on/off and make sure they only filter on the MC samples that they should be filtering on. I don't see a way around that at the moment so we'll just have to be careful. 
 
> test/EfficiencyResolutionFromMC/histos.py:
> 
> 1. one bug found [3] which looks like just a typo
>    `extra = '_'+ if ex!='' else ''` -> `extra = '_'+ex if ex!='' else ''`
> 2. I think current setup might consumes memory too much (RSS ~ 3391 MB), i.e. too large to submit the CRAB jobs... It is likely due to having too many histograms, so in principle one can reduce number of modules (EDAnalyzers) if CRAB fails.

1. Fixed, thanks. 
2. Ah, good catch. In this case do_all_track_fits should be set to False by default so that only the "normal" tracks are used (TuneP). Setting the flag to True would only be useful for special tests comparing the TeV reconstructors so, yes, it would be necessary to remove certain EDAnalyzers or cut sets to make sure the memory usage is small enough. I added a comment above the flag as a warning.

JanFSchulte commented:
Hi, 
btw, to get around some of these issues, like the gen filters, I would actually recommend adding the histos.py files for these parts of the framework into the new user interface I had added. For example it automatically adds the correct gen filter based on the sample you are processing, reducing the number of places where that has to be maintained to one. 

And I agree that the framework has memory issues, but in general I am able to run just be requesting more memory in the CRAB config.

cschnaible committed:
Finally fix gen-level filters for histos.py and remove ntuple cfg option that I added but never did anything with

cschnaible committed:
Add MCFilters_cfi for gen filtering

cschnaible commented:
Hi, I took some of the good ideas from Jan's framework interface for handling the gen filters and put them in DataMCSpectraComparison/histos.py :)

cschnaible committed:
Fixing a few issues: good data filters now applied correctly in DataMC/histos.py and genWeight now stored correctly in the n-tupler. MCSamples cleaning, too.

cschnaible committed:
Implement fixes to EfficiencyFromMCMini from Min. Clean up HLT efficiency sections, remove L1 efficiencies, add gen weighting

cschnaible committed:
Fix trigger cuts for prescaled trigger in EfficiencyFromMC; make binning in mass resolution vs mass plots settable in python cfg

khaosmos93 commented:
Hi, the core parts of the update look fine.
One thing I noticed is that the latest matched station cut is not implemented in "python/OurSelection2018_cff.py" (python/OurSelection2017_cff.py looks fine so we should synchronize).
I think PR66 is good to go if OurSelection2018_cff.py is fixed.

For the next steps, I will make PRs for following topics,
- Minor updates on nminus1effs.py (e.g. applying proper Gen filter, etc...)
- Backport some of the important updates to mini-AOD-2017 ("expectedNnumberOfMatchedStations" should be implemented as well if we want to run 2017 dataset and 94X samples with 94X release)

BenjaminRS commented:
Thanks for the check Min. Chris once you update the OurSelection2018_cff.py I think we can merge this PR.

cschnaible committed:
Update matched muon selection in OurSelection2018_cff.py

cschnaible commented:
I've made the update for the matched muon station selection in OurSelection2018_cff.py. Also, I had another commit in the pipeline that makes the mass range and binning configurable via the cfg file for ResolutionUsingMC.cc and puts the correct pt cut in for the prescaled trigger path in the EfficiencyFromMCMini submission script.
